docs/adr/0007-modelling-history.md

```diff
@@ -0,0 +1,272 @@
+particular locale. It has many editions and many revisions. At any time it will
+have a current edition -  shown on Content Publisher index - and potentially a
+live edition which is currently on GOV.UK. The live and current edition can be
+the same. Document serves as a mutable entity that tracks when a Document was
+last edited and it is expected to be a joining point for document related data
+that is not associated with a particular editions.
+
+**Edition**: An edition is a numbered version of a document that has been, or is
+expected to be, published on GOV.UK. It is associated with a current revision
+and maintains an association with all revisions that have been associated with
+it. It has a current state (e.g. draft, submitted, published)
+and has knowledge of the past states it held. It is mutable and is expected
+as a place where any edition level database constraints can be placed.
+
+**Revision**: Represents the content of a document at a particular point in
+time. It has a number to indicate which revision of the document it is and
+stores who created it. Any time a user changes an aspect of a document a new
+revision is created and the current edition is updated to reference that
+revision - a revision cannot be edited and is considered immutable. A revision
+may have a association with a collection of items, such as images, when a new
+revision is created new images are not created but a new join between them is.
+
+This modelling is represented in the following diagram:
+
+![Class diagram](0007/class-diagram.png)
+
+Note: for brevity this diagram shows a reduced selection of fields and removes
+associations with Users.
+
+The purpose of models that not described above are:
+
+**State**: This represents the state of a document at a particular point in time,
+and is associated with an edition by being the current state. This model knows
+who created the state change and when. The model will likely have an polymorphic
+association to an object with any specifics of the state (such as time for
+scheduling, reason for retiring, etc). It is currently defined as immutable
+since state changes happen atomically but may well become mutable if it needs
+to track an asynchronous change in Publishing API.
+
+**Image**: Is an aspect of the current content of a document, it has a many to
+many relationship with a revision (since an image could be associated with
+many Revisions and a single Revision may have many images). It is immutable and
+any change to an image creates a new image object. We can identify an image
+replacement from a creation due to the continuous reference to a Blob. One image
+can be defined as the lead image on a revision which is an optional association.
+
+**Blob**: Blob is a model provided by [ActiveStorage][] which is used to manage
+the underlying file on cloud or storage. It is already used in Content
+Publisher.
+
+**AssetManagerState**: This is a model defined to track the state of a file in
```

Comment:
> I think the name is confusing me here as it's a mixture of the data a Blob would store (remote ids/lookup data) and the data a State would store (is it published, etc.). Maybe it's a terminology thing: it looks like we're using 'XYZState' to mean 'data we store about XYZ'. Could you help me understand you're thinking here?

Yeah, no problem (well hopefully my explanation makes sense). So the logic for this model is it is store all the information that we need for communication between the immutable aspect of a revision and Asset Manager.

When we initially create an Image it is not stored on asset manager and has an absent state. When it is then put on the draft stack we get a file_id from asset manager and change that state and finally when live the state is changed again. Does that make sense?

a Blob is a record of the source file rather than any derivations so that wouldn't make sense to have the remote_id there.

One of the things I've realised while replying though is that this works on the assumption that there is just a single file for each Image, which isn't correct (although I assume how Content Publisher works currently) as we should at least have retina and non retina any may have other sizes too. So it might be that this concept is better considered as an AssetManagerFile which has a one to many with Image.

> * By 'editions associated with the asset are', do you mean 'the associated asset is'?

I don't think quite that. I'll try explain verbosely and then can amend the document to be more concise.

A change of state of an Edition (e.g. draft to published) should move images from draft Asset Manager to live Asset Manager. This should not change the Image model as a user has not changed that Image and thus the suitable place to change this is the AssetManagerState object

> * Why is State immutable but AssetManagerState mutable?

I probably should have flagged State as mutable in the diagram really, it's only flagged as immutable as it can be immutable if state changes synchronously.

The other aspect is that I don't expect we'll place high value in knowing all the past states and times of transitions for an Asset Manager File - that information is hopefully the domain of Asset Manager.